File: ux_validation_test.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

class UXValidationTester {
    constructor() {
        this.testsRun = 0;
        this.testsPassed = 0;
        this.errors = [];
    }

    runTest(testName, testFunction) {
        this.testsRun++;
        console.log(`\n🔍 Testing: ${testName}`);
        
        try {
            const result = testFunction();
            if (result.success) {
                this.testsPassed++;
                console.log(`✅ PASS - ${result.message}`);
            } else {
                console.log(`❌ FAIL - ${result.message}`);
                this.errors.push(`${testName}: ${result.message}`);
            }
        } catch (error) {
            console.log(`❌ FAIL - Exception: ${error.message}`);
            this.errors.push(`${testName}: Exception - ${error.message}`);
        }
    }

    loadConfig(filePath) {
        try {
            // Read file content and extract the relevant objects
            const content = fs.readFileSync(filePath, 'utf8');
            return content;
        } catch (error) {
            throw new Error(`Failed to load ${filePath}: ${error.message}`);
        }
    }

    extractKeyboardOrder(content) {
        // Extract userKeyboard array structure
        const match = content.match(/const userKeyboard = \{[\s\S]*?keyboard: \[([\s\S]*?)\],/);
        if (match) {
            return match[1].trim();
        }
        return null;
    }

    testKeyboardReordering() {
        return this.runTest("config.js userKeyboard reordered correctly", () => {
            const configContent = this.loadConfig('/app/js/config.js');
            const keyboard = this.extractKeyboardOrder(configContent);
            
            if (!keyboard) {
                return { success: false, message: "Could not extract userKeyboard from config.js" };
            }

            // Check the expected order based on review request
            const expectedOrder = [
                'user.urlShortenerMain',
                'user.hostingDomainsRedirect', 
                'user.phoneNumberLeads',
                'user.wallet, user.viewPlan',
                'user.buyPlan'
            ];

            let allFound = true;
            let foundOrder = [];

            for (const expected of expectedOrder) {
                if (keyboard.includes(expected)) {
                    foundOrder.push(expected);
                } else {
                    allFound = false;
                    break;
                }
            }

            if (allFound) {
                return { success: true, message: `Keyboard reordering correct: ${foundOrder.join(' -> ')}` };
            } else {
                return { success: false, message: `Keyboard order incorrect. Expected: ${expectedOrder.join(' -> ')}` };
            }
        });
    }

    testLanguageFileConsistency() {
        return this.runTest("All language files have consistent userKeyboard ordering", () => {
            const files = [
                '/app/js/lang/en.js',
                '/app/js/lang/fr.js', 
                '/app/js/lang/hi.js',
                '/app/js/lang/zh.js'
            ];

            const keyboards = files.map(file => {
                const content = this.loadConfig(file);
                return this.extractKeyboardOrder(content);
            });

            // Check if all keyboards have similar structure (allowing for HIDE_SMS_APP differences)
            const allConsistent = keyboards.every(kb => 
                kb && 
                kb.includes('user.urlShortenerMain') &&
                kb.includes('user.hostingDomainsRedirect') &&
                kb.includes('user.phoneNumberLeads') &&
                kb.includes('user.wallet, user.viewPlan') &&
                kb.includes('user.buyPlan')
            );

            if (allConsistent) {
                return { success: true, message: "All language files have consistent keyboard ordering" };
            } else {
                return { success: false, message: "Language files have inconsistent keyboard ordering" };
            }
        });
    }

    testTextUpdates() {
        return this.runTest("Text updates implemented correctly", () => {
            const configContent = this.loadConfig('/app/js/config.js');
            const enContent = this.loadConfig('/app/js/lang/en.js');

            const checks = [
                {
                    name: 'whatNum updated',
                    pattern: /whatNum.*That doesn't look right\. Please enter a valid number/,
                    content: configContent
                },
                {
                    name: 'walletBalanceLow includes deposit instructions',
                    pattern: /walletBalanceLow.*Tap.*My Wallet.*Deposit.*to top up/,
                    content: configContent
                },
                {
                    name: 'urlShortenerSelect added',
                    pattern: /urlShortenerSelect.*Shorten, brand, or track your links/,
                    content: configContent
                },
                {
                    name: 'phoneNumberLeads text updated in config.js',
                    pattern: /phoneNumberLeads.*Buy verified phone leads or validate your own numbers/,
                    content: configContent
                },
                {
                    name: 'phoneNumberLeads text updated in en.js',
                    pattern: /phoneNumberLeads.*Buy verified phone leads or validate your own numbers/,
                    content: enContent
                },
                {
                    name: 'askValidAmount updated in config.js',
                    pattern: /askValidAmount.*Please enter a valid amount/,
                    content: configContent
                },
                {
                    name: 'askValidAmount updated in en.js', 
                    pattern: /askValidAmount.*Please enter a valid amount/,
                    content: enContent
                },
                {
                    name: 'showDepositNgnInfo grammar fixed in config.js',
                    pattern: /wallet will be updated/,
                    content: configContent
                },
                {
                    name: 'showDepositNgnInfo grammar fixed in en.js',
                    pattern: /wallet will be updated/,
                    content: enContent
                }
            ];

            const failures = [];
            for (const check of checks) {
                if (!check.pattern.test(check.content)) {
                    failures.push(check.name);
                }
            }

            if (failures.length === 0) {
                return { success: true, message: `All ${checks.length} text updates verified` };
            } else {
                return { success: false, message: `Failed checks: ${failures.join(', ')}` };
            }
        });
    }

    testErrorMessageUpdates() {
        return this.runTest("Error messages updated to be clearer", () => {
            const configContent = this.loadConfig('/app/js/config.js');
            const enContent = this.loadConfig('/app/js/lang/en.js');

            // Check that error messages no longer have confusing text and include support info
            const errorChecks = [
                {
                    name: 'redIssueUrlBitly has clear error message',
                    pattern: /redIssueUrlBitly.*Link shortening failed.*contact.*SUPPORT_USERNAME/,
                    content: configContent
                },
                {
                    name: 'redIssueUrlCuttly has clear error message',
                    pattern: /redIssueUrlCuttly.*Link shortening failed.*contact.*SUPPORT_USERNAME/,
                    content: configContent
                },
                {
                    name: 'issueGettingPrice has clear error message in en.js',
                    pattern: /issueGettingPrice.*couldn't fetch the price.*contact.*SUPPORT_USERNAME/,
                    content: enContent
                }
            ];

            const failures = [];
            for (const check of errorChecks) {
                if (!check.pattern.test(check.content)) {
                    failures.push(check.name);
                }
            }

            if (failures.length === 0) {
                return { success: true, message: `All ${errorChecks.length} error message updates verified` };
            } else {
                return { success: false, message: `Failed checks: ${failures.join(', ')}` };
            }
        });
    }

    testTranslatedUpdates() {
        return this.runTest("Translated phoneNumberLeads text updated in all language files", () => {
            const langFiles = [
                { file: '/app/js/lang/fr.js', lang: 'French', expectedPattern: /phoneNumberLeads.*Achetez des leads téléphoniques vérifiés ou validez vos propres numéros/ },
                { file: '/app/js/lang/hi.js', lang: 'Hindi', expectedPattern: /phoneNumberLeads.*सत्यापित फ़ोन लीड खरीदें या अपने नंबर मान्य करें/ },
                { file: '/app/js/lang/zh.js', lang: 'Chinese', expectedPattern: /phoneNumberLeads.*购买经验证的电话线索或验证您自己的号码/ }
            ];

            const failures = [];
            for (const langFile of langFiles) {
                const content = this.loadConfig(langFile.file);
                if (!langFile.expectedPattern.test(content)) {
                    failures.push(langFile.lang);
                }
            }

            if (failures.length === 0) {
                return { success: true, message: "All translated phoneNumberLeads text updated correctly" };
            } else {
                return { success: false, message: `Failed translations: ${failures.join(', ')}` };
            }
        });
    }

    testSyntaxValidation() {
        return this.runTest("All files pass Node.js syntax validation", () => {
            const files = [
                '/app/js/config.js',
                '/app/js/lang/en.js',
                '/app/js/lang/fr.js',
                '/app/js/lang/hi.js',
                '/app/js/lang/zh.js'
            ];

            const { execSync } = require('child_process');
            const failures = [];

            for (const file of files) {
                try {
                    execSync(`node --check ${file}`, { stdio: 'pipe' });
                } catch (error) {
                    failures.push(path.basename(file));
                }
            }

            if (failures.length === 0) {
                return { success: true, message: `All ${files.length} files pass syntax validation` };
            } else {
                return { success: false, message: `Syntax errors in: ${failures.join(', ')}` };
            }
        });
    }

    runAllTests() {
        console.log('🚀 Starting UX Validation Testing for NomadlyBot');
        console.log('=' * 60);

        this.testKeyboardReordering();
        this.testLanguageFileConsistency();
        this.testTextUpdates();
        this.testErrorMessageUpdates(); 
        this.testTranslatedUpdates();
        this.testSyntaxValidation();

        console.log('\n' + '=' * 60);
        console.log(`📊 Test Results: ${this.testsPassed}/${this.testsRun} tests passed`);
        
        const successRate = (this.testsPassed / this.testsRun * 100);
        console.log(`📈 Success Rate: ${successRate.toFixed(1)}%`);

        if (this.errors.length > 0) {
            console.log('\n❌ Failed Tests:');
            this.errors.forEach(error => console.log(`  - ${error}`));
        }

        if (this.testsPassed === this.testsRun) {
            console.log('🎉 All UX validation tests passed!');
            return 0;
        } else {
            console.log('⚠️  Some UX validation tests failed - see details above');
            return 1;
        }
    }
}

// Run tests
const tester = new UXValidationTester();
const exitCode = tester.runAllTests();
process.exit(exitCode);

File: ux_validation_test.py
#!/usr/bin/env python3

import re
import subprocess
import sys
from pathlib import Path

class UXValidationTester:
    def __init__(self):
        self.tests_run = 0
        self.tests_passed = 0
        self.errors = []

    def run_test(self, name, test_func):
        """Run a single test and track results"""
        self.tests_run += 1
        print(f"\n🔍 Testing: {name}")
        
        try:
            success, details = test_func()
            if success:
                self.tests_passed += 1
                print(f"✅ PASS - {details}")
            else:
                print(f"❌ FAIL - {details}")
                self.errors.append(f"{name}: {details}")
            return success
        except Exception as e:
            print(f"❌ FAIL - Exception: {str(e)}")
            self.errors.append(f"{name}: Exception - {str(e)}")
            return False

    def test_keyboard_reordering_config(self):
        """Test config.js userKeyboard has correct order"""
        try:
            with open('/app/js/config.js', 'r') as f:
                content = f.read()
            
            # Extract keyboard array structure
            keyboard_match = re.search(r'const userKeyboard = \{[\s\S]*?keyboard: \[([\s\S]*?)\],', content)
            if not keyboard_match:
                return False, "Could not find userKeyboard in config.js"
            
            keyboard_content = keyboard_match.group(1)
            
            # Check the expected order based on review request
            expected_patterns = [
                r'\[user\.urlShortenerMain\]',
                r'\[user\.hostingDomainsRedirect\]', 
                r'\[user\.phoneNumberLeads\]',
                r'\[user\.wallet,\s*user\.viewPlan\]',
                r'\[user\.buyPlan\]'
            ]
            
            current_pos = 0
            found_order = []
            
            for pattern in expected_patterns:
                match = re.search(pattern, keyboard_content[current_pos:])
                if match:
                    found_order.append(pattern.replace(r'\.', '.').strip('[]r'))
                    current_pos += match.end()
                else:
                    return False, f"Missing or out of order: {pattern}"
            
            return True, f"Keyboard ordering correct: {' -> '.join(found_order)}"
        except Exception as e:
            return False, f"Error reading config.js: {str(e)}"

    def test_language_files_keyboard_consistency(self):
        """Test that all language files have proper keyboard structure"""
        try:
            files = [
                '/app/js/lang/en.js',
                '/app/js/lang/fr.js',
                '/app/js/lang/hi.js', 
                '/app/js/lang/zh.js'
            ]
            
            all_good = True
            issues = []
            
            for file_path in files:
                with open(file_path, 'r') as f:
                    content = f.read()
                
                # Check that it has the main keyboard elements in some order
                required_elements = [
                    'user.urlShortenerMain',
                    'user.hostingDomainsRedirect',
                    'user.phoneNumberLeads',
                    'user.wallet, user.viewPlan',
                    'user.buyPlan'
                ]
                
                missing = []
                for element in required_elements:
                    if element not in content:
                        missing.append(element)
                
                if missing:
                    issues.append(f"{Path(file_path).name}: missing {missing}")
                    all_good = False
            
            if all_good:
                return True, f"All {len(files)} language files have required keyboard elements"
            else:
                return False, f"Issues found: {'; '.join(issues)}"
                
        except Exception as e:
            return False, f"Error checking language files: {str(e)}"

    def test_specific_text_updates(self):
        """Test specific text updates mentioned in review request"""
        try:
            # Load config.js
            with open('/app/js/config.js', 'r') as f:
                config_content = f.read()
            
            # Load en.js 
            with open('/app/js/lang/en.js', 'r') as f:
                en_content = f.read()
            
            checks = [
                ("whatNum updated in config.js", r"whatNum.*That doesn't look right\. Please enter a valid number", config_content),
                ("walletBalanceLow includes deposit instructions", r"walletBalanceLow.*Tap.*My Wallet.*Deposit.*to top up", config_content), 
                ("urlShortenerSelect added in config.js", r"urlShortenerSelect.*Shorten, brand, or track your links", config_content),
                ("select kept as 'Please select an option'", r"select.*Please select an option", config_content),
                ("phoneNumberLeads updated in config.js", r"phoneNumberLeads.*Buy verified phone leads or validate your own numbers", config_content),
                ("phoneNumberLeads updated in en.js", r"phoneNumberLeads.*Buy verified phone leads or validate your own numbers", en_content),
                ("askValidAmount updated in config.js", r"askValidAmount.*Please enter a valid amount", config_content),
                ("askValidAmount updated in en.js", r"askValidAmount.*Please enter a valid amount", en_content),
                ("showDepositNgnInfo grammar fixed - config.js", r"wallet will be updated", config_content),
                ("showDepositNgnInfo grammar fixed - en.js", r"wallet will be updated", en_content),
            ]
            
            failed_checks = []
            for check_name, pattern, content in checks:
                if not re.search(pattern, content, re.DOTALL):
                    failed_checks.append(check_name)
            
            if not failed_checks:
                return True, f"All {len(checks)} text updates verified"
            else:
                return False, f"Failed: {', '.join(failed_checks)}"
                
        except Exception as e:
            return False, f"Error checking text updates: {str(e)}"

    def test_error_message_improvements(self):
        """Test that error messages have been improved"""
        try:
            with open('/app/js/config.js', 'r') as f:
                config_content = f.read()
            
            with open('/app/js/lang/en.js', 'r') as f:
                en_content = f.read()
            
            checks = [
                ("redIssueUrlBitly clear error", r"redIssueUrlBitly.*Link shortening failed.*contact.*SUPPORT_USERNAME", config_content),
                ("redIssueUrlCuttly clear error", r"redIssueUrlCuttly.*Link shortening failed.*contact.*SUPPORT_USERNAME", config_content),
                ("issueGettingPrice clear error in en.js", r"issueGettingPrice.*couldn't fetch.*contact.*SUPPORT_USERNAME", en_content),
            ]
            
            failed_checks = []
            for check_name, pattern, content in checks:
                if not re.search(pattern, content, re.DOTALL):
                    failed_checks.append(check_name)
            
            if not failed_checks:
                return True, f"All {len(checks)} error message improvements verified"
            else:
                return False, f"Failed: {', '.join(failed_checks)}"
                
        except Exception as e:
            return False, f"Error checking error messages: {str(e)}"

    def test_translated_phoneNumberLeads(self):
        """Test that phoneNumberLeads text is properly translated"""
        try:
            lang_files = [
                ('/app/js/lang/fr.js', 'French', r'phoneNumberLeads.*Achetez des leads téléphoniques vérifiés ou validez vos propres numéros'),
                ('/app/js/lang/hi.js', 'Hindi', r'phoneNumberLeads.*सत्यापित फ़ोन लीड खरीदें या अपने नंबर मान्य करें'),
                ('/app/js/lang/zh.js', 'Chinese', r'phoneNumberLeads.*购买经验证的电话线索或验证您自己的号码')
            ]
            
            failed_langs = []
            for file_path, lang_name, pattern in lang_files:
                with open(file_path, 'r', encoding='utf-8') as f:
                    content = f.read()
                
                if not re.search(pattern, content):
                    failed_langs.append(lang_name)
            
            if not failed_langs:
                return True, f"All {len(lang_files)} translated phoneNumberLeads texts verified"
            else:
                return False, f"Failed translations: {', '.join(failed_langs)}"
                
        except Exception as e:
            return False, f"Error checking translations: {str(e)}"

    def test_syntax_validation(self):
        """Test that all files pass Node.js syntax validation"""
        try:
            files = [
                '/app/js/config.js',
                '/app/js/lang/en.js', 
                '/app/js/lang/fr.js',
                '/app/js/lang/hi.js',
                '/app/js/lang/zh.js'
            ]
            
            failed_files = []
            for file_path in files:
                try:
                    result = subprocess.run(['node', '--check', file_path], 
                                          capture_output=True, text=True, timeout=10)
                    if result.returncode != 0:
                        failed_files.append(f"{Path(file_path).name}: {result.stderr}")
                except subprocess.TimeoutExpired:
                    failed_files.append(f"{Path(file_path).name}: timeout")
                except Exception as e:
                    failed_files.append(f"{Path(file_path).name}: {str(e)}")
            
            if not failed_files:
                return True, f"All {len(files)} files pass Node.js syntax validation"
            else:
                return False, f"Syntax errors: {'; '.join(failed_files)}"
                
        except Exception as e:
            return False, f"Error during syntax validation: {str(e)}"

def main():
    """Main test runner for UX validation"""
    tester = UXValidationTester()
    
    print("🚀 Starting NomadlyBot UX Validation Testing")
    print("=" * 60)
    
    # Run all UX validation tests
    tester.run_test("config.js userKeyboard reordered correctly", tester.test_keyboard_reordering_config)
    tester.run_test("All language files have consistent keyboard elements", tester.test_language_files_keyboard_consistency)
    tester.run_test("Specific text updates implemented", tester.test_specific_text_updates)
    tester.run_test("Error messages improved for clarity", tester.test_error_message_improvements)
    tester.run_test("phoneNumberLeads translated in all language files", tester.test_translated_phoneNumberLeads)
    tester.run_test("All files pass Node.js syntax validation", tester.test_syntax_validation)
    
    # Print summary
    print("\n" + "=" * 60)
    print(f"📊 UX Validation Results: {tester.tests_passed}/{tester.tests_run} tests passed")
    
    success_rate = (tester.tests_passed / tester.tests_run * 100) if tester.tests_run > 0 else 0
    print(f"📈 Success Rate: {success_rate:.1f}%")
    
    if tester.errors:
        print("\n❌ Failed Tests:")
        for error in tester.errors:
            print(f"  - {error}")
    
    if tester.tests_passed == tester.tests_run:
        print("🎉 All UX validation tests passed!")
        return 0
    else:
        print("⚠️  Some UX validation tests failed - see details above")
        return 1

if __name__ == "__main__":
    sys.exit(main())